CMake: Compile examples if examples component is set up

tamiko opened the pull request:
Closes #3983

tamiko commented:
@habemus-papadum Please have a look at this pull request.

tamiko commented:
All, the biggest advantage of this change in my opinion is the fact that we can phase out our kludgy "build_tests".

We now simply unconditionally compile all available example steps during the main build. If someone doesn't like that, simply disabling the examples component helps.

tamiko commented:
The only disadvantage is the fact that now configure takes a considerably longer time, e.g. 5 seconds compared to 1.7 seconds reconfigure time. Maybe we should introduce an option for that :-D

habemus-papadum commented:
@tamiko
Yeah -- this patch is what I had in mind (and seems much more robust than the quick hack i made in my fork) - thanks!  
 
The increase in configure time is probably worth addressing so as not to irritate others...

though this also potentially starts another (separate) discussion:  based on my initial reading of deal.ii's cmake infrastructure I feel like there are several things could be less engineered/simplified -- relying on CMake conventions where possible but also following the lead of llvm, qt, boost, and few other large packages in how they utilize cmake and "vend" their libs via cmake:  I think this could possibly lead to quicker incremental builds, easier integration of deal.ii into third party applications, and broader ide support, and so forth. 

Everyone has opinions about build infra, and it's certainly not my place to comment, but if you ever wanted to compare notes we could do a google hang out or something.  cheers

tamiko commented:
@habemus-papadum 

I started the CMake rewrite with the clear idea in mind that this can't be more than 1000 SLOC of CMake and we ended up with over 5000 [1] - the sad truth is simply that almost all build systems are overengineered monsters. As for the projects you cited, they have a larger codebase and consequently considerably more SLOC of CMake (llvm-clang >10k, boost doesn't use cmake but wrote its own build system, trilinos >50k).

If you can point out unnecessary complexity I am very happy to remove that. But I have the feeling that this will only end in long sad stories about why we wrote certain crazy stuff to work around cmake deficiencies. Certain things are non-negotiable, e.g.,
 - we have ~1000 SLOC for error messages alone
 - another ~1100 SLOC check for compiler features/bugs/etc.
 - I would love to replace this FEATURE_* mess by import targets, except we would require a fairly recent cmake version (to get them relatively bug free) and... they are broken.
 - parallel build of debug and release libraries adds a toll.
 - around 500 SLOC to print useful information and provide stuff like `detailed.log`.
 
As for the time consuming part - it is mostly due to gcc getting slower (we have a lot of checks for features/bugs) and that we have too many top level targets (cmake doesn't scale).

Total configure at the moment is roughly 20 seconds for me, of which
 - compiler setup 4 sec.
 - checks 7 sec.
 - feature detection and finalize 4 sec.
 - Writing stuff to disc 5 sec. (without examples 1 sec.)


[1] Roughly 6000 SLOC directly related to the library nowadays. For almost 20 external dependencies that have to be checked and set up, this is not much compared to other projects.

bangerth commented:
I don't perceive the cmake runs to be onerous. They are soo much faster than the previous autoconf system.

Does cmake support running some things in parallel? Or will it at one point? There are a number of things we can definitely do in parallel -- e.g., testing compiler flags. It would of course be nice if cmake could do this, but I suspect it's not the biggest pain point we currently have.

bangerth commented on examples/CMakeLists.txt:
```diff
@@ -46,17 +46,39 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
+      FILE(READ "${_directory}/dependency" _dependency)
+      # I will so totally go to hell for that:
+      RUN_COMMAND("
+        IF(NOT (${_dependency}))
```
Does this mean "the string is not empty"?

bangerth commented on examples/CMakeLists.txt:
```diff
@@ -46,17 +46,39 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
+    GET_FILENAME_COMPONENT(_directory ${_step} DIRECTORY)
+    SET(_set_up TRUE)
+    IF(EXISTS "${_directory}/dependency")
+      FILE(READ "${_directory}/dependency" _dependency)
```
We already have a subdir `step-XX/doc` where we list meta-info about each step. Maybe stick it in there?

Could we re-use this information in the `step-XX/CMakeLists.txt` files themselves, rather than list things twice?

tamiko replied:
If we want to hide the dependency stuff there, no problem. The file in the plain step directory is more obvious, though.

I thought about that, and opted against it. It would make the `CMakeLists.txt` files for example steps even clumsier than they already are.

bangerth replied:
Maybe, but it also duplicates information. That's almost never successful.

I'm also a fan of clean directories. If we can hide something in a subdirectory that the normal user does not need to see every day, then let's do that.

bangerth commented on examples/CMakeLists.txt:
```diff
@@ -46,17 +46,39 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
+        IF(NOT (${_dependency}))
+          MESSAGE(STATUS \"  ${_name} - dependency not satisfied\")
+          SET(_set_up FALSE)
+        ENDIF()"
```
Yes, you ought to go to hell for that. Explain to me how exactly this works? You put CMake code into a file and then include that to execute it?

I don't think I even understand what it really is you're trying to achieve here...

tamiko replied:
Good, see you in hell :-D

It is a brutal trick to evaluate the `IF(NOT <expression>) ...` condition where the expression depends on the content of external files. This is necessary because cmake does not support "self modifying" cmake script.

I find the alternative of listing *every* example step with its configuration requirements in `examples/CMakeLists.txt` less appealing.

bangerth replied:
So could we, instead of a list of dependency variables, put a little CMake script into `step-XX/doc/dependencies`? We could then just include *that*.

tamiko commented:
@bangerth
> So do I understand it right that you want to do the setup in examples/CMakeLists.txt because you don't want to do it in examples/step-XX/CMakeLists.txt? And that that is because you want to keep the examples as separate projects, so their own CMakeLists.txt files should not be changed, but you just work around this issue for the purposes of building and installing the executables?

Yes. I would prefer to not dump even more stuff into `examples/step-*/CMakeLists.txt` - they already account for 2500 lines of CMake script. This patch has +60 lines of code. Multiplying the same logic 60 times in every example step would account for roughly 600 lines of code.

tamiko commented:
@bangerth Well, let me rewrite that, addressing all of it.

bangerth commented:
I agree with not replicating the code many times over.

tamiko commented:
I will come back to this end of next week / in two weeks.

habemus-papadum commented:
Hi -- From my perspective the line count for dealii build infra seems reasonable and is not exactly what I mean by "complexity."  The one concrete timing issue that seems odd to me is rebuilding after a successful build (i.e. should do nothing) seems 4-5 times too slow but at I've yet to figure out what is causing this.  

When I reference boost, llvm, qt, I mean primarily that even though they all have arbitrarily complicated  "build jails" (i.e. if you want to make any nontrivial addition to boost, you need to understand b2, if you want to contribute to llvm, you need to learn their cmake infra, etc) -- the cmake configs they install for the end user use "modern" cmake practices.   This provides downstream users the ability to integrate these very complicated technologies with a minimum of imposition. Although currently dealii supports out of tree executables (either as documented in the manual or using "autopilot") both methods have flaws from the perspective of an "advanced" cmake user (in contrast to the approach of boost, qt and llvm (llvm is still not perfect)).

Furthermore, at first glance these "flaws" seem to stem from other design decisions (primarily, explicitly storing compiler definitions, include dirs, and  complier/linker flags in custom Cmake variables rather than as public or private target properties and then leveraging Cmake's excellent dependency tracking.  There are also some strategies that can be used when integrating the 20 or so external libs in a way that keeps things in a format where you still have cmake to do the complicated stuff e.g. using "proxy" targets for each external lib)

It's possible that I am mistaken,  but at least that those were my initial impressions as I was reading through the dealii code and thinking about how to potentially make a great project even better... :-)   I'd be happy to try to elaborate more concrete details as separate issue though I tend to think this would be more effective to discuss via a very quick skype call/hangout
  
cheers

tamiko commented:
@habemus-papadum 
Simply put, no we can't (#2576), or internally it is not worth the effort (#2543). Further, we currently guarantee cmake-2.8.8 to work.

If you want to have "modern" import targets, simply define one.

tamiko commented:
Sorry, I meant #2541

habemus-papadum commented:
thanks -- those issues clarify some of the things I was seeing, and also noting not all things are worth the effort.  

> re: If you want to have "modern" import targets, simply define one
Yup, I totally agree! this is what I usually do, and just move on as life is short.  Somehow this conversation drifted into a direction that was not quite productive -- it's basically bad form to comment on someone else's build, so please accept my apologies.

tamiko commented:
> so please accept my apologies.

No worries, there is nothing to apologize for. :-)

bangerth commented:
@tamiko -- if I see this right, then the current state is not what want merged. So I added `[WIP]` to the label to make sure nobody has the wrong ideas here. Feel free to undo if I misunderstood.

tamiko commented:
*ugh* I will come back to this :-D

tamiko commented:
@bangerth I have updated this pull request. See the following comments bundled in a review.

tamiko commented on cmake/macros/macro_evaluate_expression.cmake:
```diff
@@ -0,0 +1,31 @@
+    )
+  FILE(WRITE ${_tmp_name} "${_the_expression}")
+  INCLUDE("${_tmp_name}")
+ENDMACRO()
```
@bangerth I have thought about this macro for a bit. Given the fact that it is recommended as a workaround for a shortcoming in the cmake parser (while it is possible to substitute variables names, e.g. `${${first}-${second}}`, it is not possible to do that with expressions, e.g. `COMMAND_${suffix}()`.

While we should generally avoid the use of such idioms, they are extremely useful for some rare corner cases to drastically reduce the amount of code written.

bangerth replied:
Nice -- a run-time evaluator. Almost LISP-like.

Would it be worth (i) deleting the file again at the end, (ii) using `${CMAKE_CURRENT_BINARY_DIR}` to avoid different macros stepping on each other's files? Or is this what `${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}` does?

tamiko replied:
CMake modules usually use `${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CMakeTmp` for storage of intermediate files, none of them explicitly clean up their temporary files afterwards (CMake occasionally cleans the directory - maybe after a `try_compile` pass, or an `add_subdirectory`, who knows).

bangerth replied:
OK, no problem. I was hoping for a parallel cmake at one point, in which case these sort of things with non-unique filenames are not going to work any more. But I suspect that is so far out in the future that we need not worry about this now :-)

tamiko commented on examples/CMakeLists.txt:
```diff
@@ -46,17 +46,65 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
+        ELSE()
+          SET(_dependency_string "NOT (${_dependency_string})")
+        ENDIF()
+      ENDIF()
```
@bangerth It turns out that I can simply extract all necessary dependency information from the `CMakeLists.txt` file of the example steps directly. No clumsy dependency files.

bangerth replied:
That will work, but I'm a bit uneasy about the fragility of the approach -- you are relying on the cmake syntax used in the file, but I think we can only do this if we only test feature macros there (nothing else) and use the exact same syntax everywhere.

I still think that listing dependencies in some standard format somewhere in a separate file may be the better approach, but I'm not going to push for it. What do others think?

tamiko replied:
I would like to avoid duplication of this information. That's why I went for extracting the information from the CMakeLists.txt files directly.

Putting dependency information in an extra file is also not terribly appealing because it would (a) either imply increasing the logic in the CMakeListst.txt files of the example steps significantly, or (b) duplicating the information.

bangerth replied:
Alright, no problem.

tamiko commented on examples/CMakeLists.txt:
```diff
@@ -46,17 +46,63 @@ IF(DEAL_II_COMPONENT_EXAMPLES)
+      EVALUATE_EXPRESSION("
+        IF(${_dependency_string})
+          SET(_setup TRUE)
+        ENDIF()")
```
The price we have to pay is this call to the `EVALUATE_EXPRESSION` macro.
The problem is that writing `IF(${_dependency_string})` directly in the cmake sources here means something subtly different - it does not evaluate the contents of `${_dependency_string}`.

bangerth replied:
Such a crazy construct given that it sets a variable completely outside the evaluation expression :-) Very nice!

tamiko replied:
That way it is still understandable I hope.

tamiko commented on examples/step-17/CMakeLists.txt:
```diff
@@ -38,7 +38,7 @@ ENDIF ()
 # Are all dependencies fulfilled?
 #
-IF(NOT DEAL_II_WITH_PETSC OR DEAL_II_PETSC_WITH_COMPLEX)
+IF(NOT DEAL_II_WITH_PETSC OR DEAL_II_PETSC_WITH_COMPLEX) # keep in one line
```
@bangerth Further, this line has to have the exact format
```
IF(...) # keep in one line
```
But I guess this is not a big issue - after all we all such code tends to get copied from other example steps.

tamiko commented:
@bangerth 
> @tamiko -- if I see this right, then the current state is not what want merged. So I added [WIP] to the label to make sure nobody has the wrong ideas here. Feel free to undo if I misunderstood.

I added the blue "WIP" label to indicate that I do not want the old state of the pull request to get merged.

bangerth commented on cmake/setup_cached_variables.cmake:
```diff
@@ -345,6 +346,12 @@ OPTION(DEAL_II_DOXYGEN_USE_MATHJAX
 
+OPTION(DEAL_II_COMPILE_EXAMPLES
+  "If set to ON, all configurable example executables will be built and installed as well. If set to OFF, the examples component only installs the source code of example steps."
+  OFF
```
I thought we had agreed that doing so is a useful thing. Maybe make the default `ON`?

tamiko replied:
Enabling >100 additional top level targets comes with a significant performance hit (not in the configure stage, but afterwards during makefile generation). I will do some benchmarking on a slow nfs file system to cover a frequent worst case scenario.

bangerth replied:
But it also comes with a price if we don't enable it by default: (i) nobody is ever going to discover the correct flag, (ii) because nobody uses it, we incur bitrot.

If the additional cost is not too large, I would favor enabling it unconditionally.

tamiko replied:
See timing results. I have changed the default to `ON`.

bangerth commented:
Looks good to me. I would still like to enable it by default, if that is possible.

I am worried that duplicating the build instructions will tie our hands in the future in having tutorial programs that may be set up differently than the ones we currently have. But that is a separate issue and maybe we need to address that when the time comes.

tamiko commented:
@bangerth I have no problem enabling the compilation of examples by default. 

Reconfigure times for
 - nfs: 13.8s versus 14.9s (without mpi)
 - ssd: 0.629s versus 0.978s (without mpi)
 - ssd: 1.434s versus 5.025s (with all external dependencies)

That's interesting. It means that the big hit in target generation is due to our gigantic link interface the moment trilinos is set up... But I guess whoever configures deal.II with everything can also configure with `-DEXAMPLES=OFF`.

tjhei commented:
> I would like to avoid duplication of this information. That's why I went for extracting the information from the CMakeLists.txt files directly.

To give a second opinion: I also think having something that fragile is a bad idea. It also requires that the "# keep in one line" comment has to stay in this place exactly, right? Is there no mechanism inside cmake to configure a project and ignore if it fails?

tamiko commented:
@tjhei
I avoided configuring example (subprojects) from the main project. Let's not do that. (It is ugly and not guaranteed to work - subproject configuration depends on information not yet available during main-project configuration).

If the extraction seems to be a bit too fragile, we can also make it more robust.

I want to point out that we introduce example steps also very infrequently. So a bit of special treatment of that one dependency line isn't going to be a gigantic maintenance burden. I fail to see how this is such a bad idea - I considered it to be a meaningful compromise between special treatment of one line in the CMakeLists.txt files and otherwise introducing more files, duplicating information, and (possibly) adding more duplicated logic to over 50 step configurations.

And it was possible to do this with a net-increase of only 80 lines of code.

Further, I changed it to the current version because the extra files duplicating dependency information was (rightfully) criticised. Now, following @bangerth's suggestion, extracting the information from the CMakeLists.txt files is not OK either. I am frankly a bit lost what to do.

Just tell me how you want this feature to be implemented - I really do not care too much. (Preferably something that doesn't add yet another 3h of work to this pull request.)

bangerth commented:
I think the approach you took is reasonable. You did all of the work, so we should give you some leeway to also decide how to actually do it. And you're right, it's not like we add tutorials often, so we can deal with some tight dependencies.

bangerth commented:
Like I said before, I'm ok with merging as is now that the default is `ON`.

tjhei commented:
> I am frankly a bit lost what to do.

I was not pushing to block this PR, sorry if it sounded that way. I just thought a second opinion was helpful in this case. I don't care too much, so lets go ahead and keep it. Is there any risk with merging this before 8.5?

tamiko commented:
@bangerth @tjhei Please excuse me - I was primarily surprised that this wasn't the little 30 minute pull request I thought it would be.

@tjhei The pull request should be safe to merge.

Can someone with an IDE check whether all example steps show up nicely?

habemus-papadum commented:
@tamiko Sorry for not contributing much to the discussion.  But I will try out the Clion IDE and  tools using clang's libtooling, and let you know how that looks.  best, nehal

tjhei commented:
Can we disable this on windows and/or when linking statically? This is terribly slow on windows right now.

habemus-papadum commented:
Clion works without issues -- coulld clone, open, and then immediately debug step-1 with no extra steps.  On a vanilla mac box with a few but not all external dep's -- the system correctly disabled about 8 examples, and the rest show up in the IDE ready to be debugged immediately.  I didn't notices any significant slowdown in during cmake generation.  the link phase is terribly slow but that seems to be due to other design decisions.

tamiko commented:
@tjhei Unconditionally disabled for Windows now.
